chips/stm32f4xx/src/dma/dma2.rs

```diff
@@ -0,0 +1,129 @@
+use kernel::platform::chip::ClockInterface;
+use kernel::utilities::StaticRef;
+
+use crate::nvic;
+use crate::rcc;
+use crate::usart;
+
+use super::{
+    ChannelId, Direction, DmaClock, DmaRegisters, FifoSize, Msize, Psize, Size, Stream, StreamId,
+    StreamPeripheral, StreamServer, TransferMode,
+};
+
+const DMA2_BASE: StaticRef<DmaRegisters> =
+    unsafe { StaticRef::new(0x40026400 as *const DmaRegisters) };
+
+/// List of peripherals managed by DMA2
+#[allow(non_camel_case_types, non_snake_case)]
+#[derive(Copy, Clone, PartialEq, Debug)]
+pub enum Dma2Peripheral {
+    USART1_TX,
+    USART1_RX,
+}
+
+impl Dma2Peripheral {
+    // Returns the IRQ number of the stream associated with the peripheral. Used
+    // to enable interrupt on the NVIC.
+    pub fn get_stream_irqn(&self) -> u32 {
+        match self {
+            Dma2Peripheral::USART1_TX => nvic::DMA2_Stream7,
+            Dma2Peripheral::USART1_RX => nvic::DMA2_Stream5, // could also be Stream 2, chosen arbitrarily
+        }
+    }
+
+    pub fn get_stream_idx<'a>(&self) -> usize {
+        usize::from(StreamId::from(*self) as u8)
+    }
+}
+
+impl From<Dma2Peripheral> for StreamId {
+    fn from(pid: Dma2Peripheral) -> StreamId {
+        match pid {
+            Dma2Peripheral::USART1_TX => StreamId::Stream7,
+            Dma2Peripheral::USART1_RX => StreamId::Stream5,
+        }
+    }
+}
+
+impl StreamPeripheral for Dma2Peripheral {
+    fn transfer_mode(&self) -> TransferMode {
+        TransferMode::Fifo(FifoSize::Full)
+    }
+
+    fn data_width(&self) -> (Msize, Psize) {
+        (Msize(Size::Byte), Psize(Size::Byte))
+    }
+
+    fn channel_id(&self) -> ChannelId {
+        match self {
+            // USART1_TX Stream 7, Channel 4
+            Dma2Peripheral::USART1_TX => ChannelId::Channel4,
+            // USART1_RX Stream 5, Channel 4
+            Dma2Peripheral::USART1_RX => ChannelId::Channel4,
+        }
+    }
+
+    fn direction(&self) -> Direction {
+        match self {
+            Dma2Peripheral::USART1_TX => Direction::MemoryToPeripheral,
+            Dma2Peripheral::USART1_RX => Direction::PeripheralToMemory,
+        }
+    }
+
+    fn address(&self) -> u32 {
+        match self {
+            Dma2Peripheral::USART1_TX => usart::get_address_dr(usart::USART1_BASE),
+            Dma2Peripheral::USART1_RX => usart::get_address_dr(usart::USART1_BASE),
+        }
+    }
+}
+
+pub fn new_dma2_stream<'a>(dma: &'a Dma2) -> [Stream<'a, Dma2<'a>>; 8] {
+    [
+        Stream::new(StreamId::Stream0, dma),
+        Stream::new(StreamId::Stream1, dma),
+        Stream::new(StreamId::Stream2, dma),
+        Stream::new(StreamId::Stream3, dma),
+        Stream::new(StreamId::Stream4, dma),
+        Stream::new(StreamId::Stream5, dma),
+        Stream::new(StreamId::Stream6, dma),
+        Stream::new(StreamId::Stream7, dma),
+    ]
+}
+
+pub struct Dma2<'a> {
+    registers: StaticRef<DmaRegisters>,
+    clock: DmaClock<'a>,
+}
```

Comment:
Can you elaborate a bit on the decision to make `Dma1` and `Dma2` different types, rather than having a single `Dma` struct with multiple instantiations? I understand this design is necessary for `StreamServer` to have `Peripheral` as an associated type rather than a generic type, but it does lead to some code duplication between `Dma1` and `Dma2`. Does using a generic parameter for `StreamServer` cause issues with propagation of generic parameters throughout the code base?

Reply:
I did that mostly because I wanted the streams from the different Dma's to be of different types. I realize now that parameterizing the Stream over the Peripheral would also have worked.

Another benefit of the current approach is that it is impossible to use a Dma with the wrong Peripheral type.
Because the Stream is associated with the Dma on a type-level basis.
If the Stream were made generic over the Peripheral, i.e. `Stream<Dma1Peripheral>`, one might accidentally create such a stream for the wrong Dma instance.

It's definitely a tradeoff between having code duplication vs. more checks at compile time.
With the current implementation the Dma is propagated to the types of Usart & SPI, so usart1 is a Usart<Dma2>.
In theory, we could simply have a single DmaPeripheral enum, that contains all peripherals, with a single, non-generic Stream struct implementation.
This way we wouldn't have any code duplication and both Usart and Spi would no longer have to be generic over the Dma or its peripheral.
The disadvantage then would be, that one could create a Usart instance for usart1 and accidentally pass it a stream of dma1, instead of dma2. Currently, this is impossible, as they are of different types.

We could have these checks at runtime, with the Peripheral reporting which Dma it belongs to and the system panicking if it's set up incorrectly, like I have added for checking whether the peripheral is added to the right stream.

So we basically have three options here:
1. Keep Stream<DmaX>
2. Change to Stream<DmaXPeripheral>
3. Remove Generics from Stream & add runtime checks

In order of most code duplication & compile-time safety to least.
It's probably best if you decide what fits best with the current design-philosophy of Tock OS :)

Reply:
Static checks are good, and the code duplication here looks manageable (i.e. it's pretty formulaic and unlikely to need to change much if at all).

I think it makes sense to combine these into a single `dma.rs` file, and move this thread to a comment in the source explaining what static properties this structure is designed to enforce.

Reply:
+1 to Brad's suggestion; thanks for explaining the reasoning here!

Reply:
:+1: 
I'll move everything into the shared dma module then and add a comment.

Reply:
Done